### PR TITLE
Fix wrong averaging of standard deviations

### LINKED ISSUE #307: wrong averaging of AR standard deviations
In `train_gv_AR` we take the average over the AR params - including the standard deviation. However, this is wrong: you have to average the covariances (or even take care of the size of the samples: https://stats.stackexchange.com/q/25848)

https://github.com/MESMER-group/mesmer/blob/acd80e40ef12ab584dff438fa56fd54f6ba43b51/mesmer/calibrate_mesmer/train_gv.py#L195-L208

These are used here:

https://github.com/MESMER-group/mesmer/blob/acd80e40ef12ab584dff438fa56fd54f6ba43b51/mesmer/create_emulations/create_emus_gv.py#L176

_Originally posted by @mathause in https://github.com/MESMER-group/mesmer/issues/306#issuecomment-1736922860_

### PR DESCRIPTION
**Pull request by veni-vidi-vici-dormivi**
First step towards #307, still need to think about weighting
 - [x] Tests added
 - [x] Fully documented, including `CHANGELOG.rst`

### CODE REVIEW

**Comment by veni-vidi-vici-dormivi** [comment]
Omg this was more work than I thought but I think it's done now. Had to change some of the old code to get it to run, though the only numerical change should still be the averaging, as @mathause already switched to passing the variance instead of the std in the old codepath too. Interestingly, the old implementation showed that `ens_dim` must also have coordinates for my implementation to work cleanly.

File: mesmer/calibrate_mesmer/train_lv.py [hunk 1]
```diff
@@ -231,13 +231,19 @@ def train_lv_AR1_sci(params_lv, targs, y, wgt_scen_eq, aux, cfg):
 
         # create temporary DataArray
         dims = ("run", "time", "cell")
-        data = [xr.DataArray(data, dims=dims) for data in targ.values()]
+        data = [
+            xr.DataArray(
+                data,
+                dims=dims,
```
Comment (mathause) [inline mesmer/calibrate_mesmer/train_lv.py:237]:
```suggestion
                dims=dims
```

File: mesmer/stats/_auto_regression.py [hunk 2]
```diff
@@ -554,6 +577,8 @@ def fit_auto_regression(data, dim, lags):
     if np.ndim(lags) == 0:
         lags = np.arange(lags) + 1
 
+    # return intercept, coeffs, variance, lags, nobs
+
```
Comment (mathause) [inline mesmer/stats/_auto_regression.py:581]:
```suggestion
```

File: tests/unit/test_auto_regression_scen_ens.py [hunk 3]
```diff
@@ -61,23 +62,32 @@ def test_select_ar_order_scen_ens_no_ens_dim():
     xr.testing.assert_equal(result, expected)
 
 
-def test_fit_auto_regression_scen_ens_one_scen():
+@pytest.mark.parametrize("da_std", [1, 0.1, 0.5])
+def test_fit_auto_regression_scen_ens_one_scen(da_std):
```
Comment (mathause) [inline tests/unit/test_auto_regression_scen_ens.py:66]:
```suggestion
@pytest.mark.parametrize("std", [1, 0.1, 0.5])
def test_fit_auto_regression_scen_ens_one_scen(std):
```

File: tests/unit/test_auto_regression_scen_ens.py [hunk 4]
```diff
@@ -61,23 +62,32 @@ def test_select_ar_order_scen_ens_no_ens_dim():
...
 
-    da = generate_ar_samples([1, 0.5, 0.3, 0.4], n_timesteps=100, n_ens=4)
+    n_timesteps = 100
+    da = generate_ar_samples(
+        [1, 0.5, 0.3, 0.4], da_std, n_timesteps=n_timesteps, n_ens=4
```
Comment (mathause) [inline tests/unit/test_auto_regression_scen_ens.py:70]:
```suggestion
        [1, 0.5, 0.3, 0.4], std, n_timesteps=n_timesteps, n_ens=4
```

File: tests/unit/test_auto_regression_scen_ens.py [hunk 5]
```diff
@@ -61,23 +62,32 @@ def test_select_ar_order_scen_ens_no_ens_dim():
...
+    expected["variance"] = sum(expected.variance * (expected.nobs - 1)) / sum(
+        expected.nobs - 1
+    )
+    expected["coeffs"] = expected.coeffs.mean("ens")
+    expected["intercept"] = expected.intercept.mean("ens")
+    expected = expected.drop_vars(["nobs", "ens"])
 
-    xr.testing.assert_equal(result, expected)
+    xr.testing.assert_allclose(result, expected)
+    np.testing.assert_allclose(np.sqrt(result.variance), da_std, rtol=1e-1)
```
Comment (mathause) [inline tests/unit/test_auto_regression_scen_ens.py:86]:
```suggestion
    np.testing.assert_allclose(np.sqrt(result.variance), std, rtol=1e-1)
```

File: tests/unit/test_auto_regression_scen_ens.py [hunk 6]
```diff
@@ -89,22 +99,30 @@ def test_fit_auto_regression_scen_ens_multi_scen():
     da = da.stack(scen_ens=("scen", "ens")).dropna("scen_ens")
     expected = mesmer.stats.fit_auto_regression(da, dim="time", lags=3)
     expected = expected.unstack()
-    expected["standard_deviation"] = np.sqrt(expected.variance)
-    expected = expected.mean("ens").mean("scen")
+    expected_coeffs = expected.coeffs.mean("ens").mean("scen")
+    expected_intercept = expected.intercept.mean("ens").mean("scen")
+    ens_variance = (expected.variance * (expected.nobs)).sum(dim="ens") / (
+        expected.nobs
+    ).sum(dim="ens")
```
Comment (mathause) [inline tests/unit/test_auto_regression_scen_ens.py:106]:
```suggestion
    ens_variance = expected.variance.weighted(expected.nobs).mean("ens")
```

File: mesmer/stats/_auto_regression.py [hunk 7]
```diff
@@ -81,26 +83,46 @@ def _fit_auto_regression_scen_ens(*objs, dim, ens_dim, lags):
...
+        return ar_params
+
     ar_params_scen = list()
-    for obj in objs:
-        ar_params = fit_auto_regression(obj, dim=dim, lags=int(lags))
+    n_ens = np.ones(len(objs))
 
-        # BUG/ TODO: fix for v1, see https://github.com/MESMER-group/mesmer/issues/307
-        ar_params["standard_deviation"] = np.sqrt(ar_params.variance)
+    for o, obj in enumerate(objs):
```
Comment (mathause) [inline mesmer/stats/_auto_regression.py:103]:
```suggestion
    for i, obj in enumerate(objs):
```

(less confusing than `o`)

File: mesmer/stats/_auto_regression.py [hunk 8]
```diff
@@ -81,26 +83,46 @@ def _fit_auto_regression_scen_ens(*objs, dim, ens_dim, lags):
...
+        ar_params = fit_auto_regression(obj, dim=dim, lags=int(lags))
 
         if ens_dim in ar_params.dims:
-            ar_params = ar_params.mean(ens_dim)
-
+            # mean over ensemble members
+            ar_params = _avg_ar_params(ar_params, ens_dim, ar_params.nobs)
+            n_ens[o] = obj[ens_dim].size
```
Comment (mathause) [inline mesmer/stats/_auto_regression.py:109]:
```suggestion
            n_ens[i] = obj[ens_dim].size
```

File: tests/integration/test_draw_realisations_from_bundle.py [hunk 9]
```diff
@@ -38,7 +38,7 @@ def test_make_realisations(
     n_realisations,
     outname,
     test_data_root_dir,
-    update_expected_files,
+    update_expected_files=False,
```
Comment (mathause) [inline tests/integration/test_draw_realisations_from_bundle.py:41]:
```suggestion
    update_expected_files,
```

**Review by mathause — COMMENTED** [review]
Some trivial comments for now. I think you can use `var.weighted(nobs).mean(dim)` for the averaging.

I would like to discuss this with Lukas but that could be difficult given our time frame.

**Comment by veni-vidi-vici-dormivi** [comment]
I would want to merge this and open a new PR for implementing the option for different weighting.

**Review by mathause — APPROVED** [review]
Ok as discussed let's merge this with the intention to offer different averaging options in another PR and re-visit the statistics.

**Pull request merged**